### PR TITLE
stream: fix infinite select() polling on osx

### PR DESCRIPTION
**Pull request by indutny**
Interruption FD buffer should be emptied after write, otherwise select()
loop will poll indefinitely.

/cc @bnoordhuis

### CODE REVIEW

**Comment by saghul** [comment]
Sorry if it's a dumb question, but, can `read` be interrupted after it has read some data? That is, is it possible that r is > 0 and errno is EINTR because there is data to be read?

**Comment by indutny** [comment]
@saghul no, it can't:

```
If successful, the number of bytes actually read is returned.
Upon reading end-of-file, zero is returned.
Otherwise, a -1 is returned and the global variable errno is set to indicate the error.
```

**Comment by saghul** [comment]
Thanks for clarifying it, Fedor.

File: src/unix/stream.c [hunk 1]
```diff
@@ -128,10 +128,10 @@ void uv__stream_init(uv_loop_t* loop,
...
   uv_stream_t* stream;
   uv__stream_select_t* s;
-  fd_set read;
-  fd_set write;
-  fd_set error;
-  struct timeval timeout;
+  char buf[1024];
+  fd_set sread;
+  fd_set swrite;
+  fd_set serror;
```
Comment (bnoordhuis) [inline src/unix/stream.c:134]: Not really related to this pull request but do you expect the error set to ever have anything in it? I'm not sure about OS X, but on Linux it's only useful for out-of-band data on TCP sockets and monitoring sysfs changes.

File: src/unix/stream.c [hunk 2]
```diff
@@ -178,13 +177,33 @@ void uv__stream_osx_select(void* arg) {
     if (r == 0)
       continue;
 
+    /* Empty socketpair's buffer in case of interruption */
+    if (FD_ISSET(s->int_fd, &sread))
+      while (1) {
+        r = read(s->int_fd, buf, sizeof(buf));
+
+        if (r == sizeof(buf))
```
Comment (bnoordhuis) [inline src/unix/stream.c:185]: What if you read less than 1024 bytes?

Comment (bnoordhuis) [inline src/unix/stream.c:185]: Oh wait, that's caught a few lines below (r != -1). Seems like you could coalesce those.

Comment (indutny) [inline src/unix/stream.c:185]: I've copied this from async.c

Comment (bnoordhuis) [inline src/unix/stream.c:185]: Oh, okay. Still - don't trust the guy who wrote that.

**Comment by bnoordhuis** [comment]
LGTM

**Comment by indutny** [comment]
Thanks, landed in ba83510

**Pull request closed without merging**